Speed up element data

Pirulax opened the pull request:
This PR improves element data perfomrnace by 22%(at least for now server-side), I'll implement it client-side as well tomorrow, hopefully.

**Renamed the branch so i need a new PR, the old one was #907**

Benchmarks can be found @ #907 

[Test resource](https://github.com/multitheftauto/mtasa-blue/files/3136547/tester.zip)

Pirulax committed:
Optimized setElementData on the server-side, but its throwing errors as of now

Pirulax committed:
Implement CFastHashMap to server-side, fix errors

Pirulax committed:
Remove MapInsert function from CFastHashMap

Pirulax committed:
fix license in CCustomData.cpp

Pirulax committed:
Fix indentation

Pirulax committed:
Fix indentation

Pirulax committed:
fix license in CCustomData.cpp finally

Pirulax committed:
Change OldVariable to oldVariable as suggested by Qaisjp

Pirulax committed:
Implement faster element data on client-side as well(there are still some problems, trying to fix them

Pirulax committed:
remove unnecesarry conversion from SString to SStringX

Pirulax committed:
Change SString strName to const char* szName because of crashes(it was const char* originally)

Pirulax committed:
Update

qaisjp commented on Client/mods/deathmatch/logic/CCustomData.cpp:
```diff
@@ -14,35 +14,40 @@
 
 SCustomData* CCustomData::Get(const char* szName)
 {
-    assert(szName);
```
Why was this (and other assertions) removed?

Pirulax replied:
oh.
I removed them because originally I repalced everything with SString, because it was converted to SString anyways and there was a conversion from SString to const char* when CLuaElementDefs called CStaticFuncDefs::SetED.
I think I'll just put SString back

qaisjp commented on Client/mods/deathmatch/logic/CCustomData.cpp:
```diff
@@ -14,35 +14,40 @@
-void CCustomData::Set(const char* szName, const CLuaArgument& Variable, bool bSynchronized)
+bool CCustomData::Set(const char* szName, const CLuaArgument& Variable, bool bSynchronized, CLuaArgument* pOldVariable)
 {
-    assert(szName);
```
https://github.com/multitheftauto/mtasa-blue/pull/912/files#r281029852

qaisjp commented on Client/mods/deathmatch/logic/CCustomData.cpp:
```diff
@@ -14,35 +14,40 @@
-
     // Grab the item with the given name
-    SCustomData* pData = Get(szName);
+    SCustomData* pData = MapFind(m_Data, szName);
```
This should work with the original `Get(szName)`

qaisjp commented on Client/mods/deathmatch/logic/CCustomData.cpp:
```diff
@@ -14,35 +14,40 @@
     if (it != m_Data.end())
         return &it->second;
 
     return NULL;
```
Maybe just replace the content of `Get` with a call with `MapFind`?

Pirulax replied:
We could but that'd be just a pointless function call, and we can to squeeze out as much performance as we can, and I think this isn't premature optimization(yet)

qaisjp replied:
I don't think there will be a real difference

Pirulax replied:
Isn't it converting from SString to const char* and then back?
Or it's optimized by the compiler already?

qaisjp replied:
99% chance optimised away. If in doubt, measure the difference

qaisjp commented on Server/mods/deathmatch/logic/CCustomData.cpp:
```diff
@@ -60,7 +50,7 @@ void CCustomData::UpdateSynced(const char* szName, const CLuaArgument& Variable,
     if (bSynchronized)
     {
-        SCustomData* pDataSynced = GetSynced(szName);
+        SCustomData* pDataSynced = MapFind(m_SyncedData, szName);
```
This should use GetSynced instead of MapFind

qaisjp commented on Server/mods/deathmatch/logic/CCustomData.cpp:
```diff
@@ -75,23 +65,30 @@ void CCustomData::UpdateSynced(const char* szName, const CLuaArgument& Variable,
 
     // Grab the item with the given name
-    SCustomData* pData = Get(szName);
+    SCustomData* pData = MapFind(m_Data, szName);
```
Get should still be used here. That way if the underlying data structure changes, fewer references to m_Data need to be changed

qaisjp commented on Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:
```diff
@@ -986,19 +986,14 @@ bool CStaticFunctionDefinitions::SetElementID(CClientEntity& Entity, const char*
 
 bool CStaticFunctionDefinitions::SetElementData(CClientEntity& Entity, const char* szName, CLuaArgument& Variable, bool bSynchronize)
 {
-    assert(szName);
```
https://github.com/multitheftauto/mtasa-blue/pull/912/files#r281029852

qaisjp commented on Client/mods/deathmatch/logic/CClientEntity.cpp:
```diff
@@ -458,33 +458,29 @@ bool CClientEntity::GetCustomDataBool(const char* szName, bool& bOut, bool bInhe
-    SCustomData* pData = m_pCustomData->Get(szName);
-    if (pData)
+    CLuaArgument OldVariable;
+    if (m_pCustomData->Set(szName, Variable, bSynchronized, &OldVariable))
```
Does it work correctly if there is no previous element data?

Pirulax replied:
It does, since the original one worked like this as well, it only set a value to CLuaArgument if there was something, otherwise an empty CLuaArgument is == nil

qaisjp commented on Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp:
```diff
@@ -1777,9 +1777,9 @@ int CLuaElementDefs::SetElementData(lua_State* luaVM)
             if (strKey.length() > MAX_CUSTOMDATA_NAME_LENGTH)
-            {
-                // Warn and truncate if key is too long
+            { 
```
```suggestion
            {
```

qaisjp reviewed (changes requested):
Thanks for the PR

Pirulax committed:
changes made + inlined Map functions, hope i didnt broke anything

Pirulax committed:
changed comment in CRPCFunctions.h to describe the reason behind ProcessPacket - FIXME thingy

Pirulax commented:
I just feel like m_SyncedData is so premature optimization... 
Because think about it.. Most of the time elementdata is synced, so you have 2x the memory usage, and you dont gain that much(if at all), but you lose when updating the element data, because you need to update it in the syncedData map as well. Maybe an std::referENCe_wrapper will help, but even then you have another SString as the key..
Maybe we can just remove the m_SyncedData entirely? 

*Edit: yea, its referENCe wrapper, not refernece

qaisjp commented:
`Maybe we can just remove the m_SyncedData entirely?`

Yeah a few months ago (when we were discussing this in #mta.dev) I thought this looked unnecessary too. I think I looked into removing it, and then realised it was actually necessary. I think? I don't really remember.

prnxdev commented:
@qaisjp what is #mta.dev?

patrikjuvonen commented:
> @qaisjp what is #mta.dev?

#mta.dev is MTA's old IRC channel for development discussions, nowadays it is bridged with our #development Discord channel.

Pirulax commented:
Its used twice only, at least thats what VS19 thinks, i hope thats the reality.

Pirulax commented:
So, whats the plan @qaisjp?

qaisjp committed:
Update Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp

qaisjp commented on Client/mods/deathmatch/logic/CCustomData.cpp:
```diff
@@ -14,35 +14,40 @@
 {
-    assert(szName);
-
     // Grab the item with the given name
```
```suggestion
    assert(szName);

    // Grab the item with the given name
```

qaisjp commented on Client/mods/deathmatch/logic/CClientEntity.cpp:
```diff
@@ -458,33 +458,29 @@ bool CClientEntity::GetCustomDataBool(const char* szName, bool& bOut, bool bInhe
+bool CClientEntity::SetCustomData(const char* szName, const CLuaArgument& Variable, bool bSynchronized)
 {
-    assert(szName);
     if (strlen(szName) > MAX_CUSTOMDATA_NAME_LENGTH)
```
```suggestion
    assert(szName);
    if (strlen(szName) > MAX_CUSTOMDATA_NAME_LENGTH)
```

qaisjp commented on Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:
```diff
@@ -986,19 +986,14 @@ bool CStaticFunctionDefinitions::SetElementID(CClientEntity& Entity, const char*
-    bool          bIsSynced;
-    CLuaArgument* pCurrentVariable = Entity.GetCustomData(szName, false, &bIsSynced);
-    if (!pCurrentVariable || Variable != *pCurrentVariable || bIsSynced != bSynchronize)
+    if (Entity.SetCustomData(szName, Variable, bSynchronize))
```
```suggestion
    assert(szName);
    if (Entity.SetCustomData(szName, Variable, bSynchronize))

```

qaisjp committed:
Apply suggestions from code review

qaisjp commented on Client/mods/deathmatch/logic/CCustomData.cpp:
```diff
@@ -14,35 +14,42 @@
-    assert(szName);
-
-    std::map<std::string, SCustomData>::iterator it = m_Data.find(szName);
+    CFastHashMap<SString, SCustomData>::iterator it = m_Data.find(szName);
```
```suggestion
    assert(szName);
    CFastHashMap<SString, SCustomData>::iterator it = m_Data.find(szName);
```

qaisjp committed:
Update Client/mods/deathmatch/logic/CCustomData.cpp

qaisjp committed:
delete make_uname.bat

Pirulax commented:
Okay, so what are the plans with this PR?

qaisjp commented:
you'll need to resolve conflicts. or redo the code.

you can reuse the same PR (don't create another one) - just force push to this branch.

Pirulax commented:
Whats up with the PR that Tederis made?

edit: i see, it was merged.

Pirulax commented:
So in the end, is m_SyncedData necessary?

qaisjp commented:
Whatever you did broke the PR, so I've undone your changes.

Pirulax committed:
Implemeting faster element data search, also optimizing existing code

Pirulax committed:
its fucked up, ill rewrite it once again, with another approach

Pirulax committed:
server/CLuaArgument: small refactor, and add a move semantics support

Pirulax committed:
server/CLuaArgument: small refactor, and add a move semantics support

Pirulax committed:
rewrite existing set/get/delete/remove/ElementData, also a few refactors to edata. sub. functions

Pirulax committed:
fix CStaticFuncDefs::SyncElementData, refactor CPlayerManager::Broadcasting functions

fix CStaticFuncDefs::SyncElementData settings the wrong ElementDataUsage out count.

Pirulax committed:
Delete duplicated code(git went weird, and copied the file into itself

Pirulax committed:
Refactor CGame::Packet_CustomData

Pirulax committed:
*/CLuaArguments.*

Add PushArgument(CLuaArgument&&) [reason: obviously, move semantics]. Also remove duplicate code added by git

Pirulax committed:
Rewrite CCustomData

Pirulax committed:
Small refactor to set/getElementData function

Pirulax committed:
Remove duplicate code(added by git probably)

Pirulax committed:
Small refactor of CStaticFuncDefs.cpp(2 lines)

Pirulax committed:
Fix CLuaArgument constructor def and declaration mismatch. Fix typo

Pirulax committed:
Add move semantics support to CLuaArguments::PushArgument

Pirulax committed:
Fix compiler error

Pirulax commented:
> Whatever you did broke the PR, so I've undone your changes.

Nah, its okay, i purposefully `git reset --head HARD`-ed it, because the old one didnt suit this anyways, and with that knowledge i improved it... "Improved it".. More like i had some fun, because tbh, the differences are quite.. small, to say the least. But try it yourself, I dont think its really worth it merging this PR.

Try the test resource from #907.
Dont look at the old results, they're now invalid, every set function doubled, probably due to that little branching(`argStream.NextIsBoolean()`) that was added. But it doubling the call time seems very bad to me.. I even tried the 1.5.7 nightly x64 server, it was a little slower(~10%), but was still twice slower than the older version with just reading a bool value. Is it possible thats causing such a huge difference?

qaisjp commented:
I haven't looked at your recent code yet. You can try converting the function to the new lua arg parser, as that apparently speeds stuff up.

Pirulax commented:
Oh, btw dont belive these failing.. 

Actually, do, because the client really is, because im a lazy fuck, but the server works fine, just pull the branch, and test it out, I'd be glad, so we can see if there really isnt much of a difference, other than way smaller memory consumption(i made a m_broadcasted, and a m_localOrSubbed maps, so in theory now it should use like 3/4 less memory, because most of the data that is set is synced, and if its synced used to be in both  maps, with 2 SCustomData's)

Pirulax commented:
> I haven't looked at your recent code yet. You can try converting the function to the new lua arg parser, as that apparently speeds stuff up.

Uh oh, where can I find that magic?

qaisjp commented:
Take a look at CLuaBcryptDefs - you'll need to use `ArgumentParserWarn<MethodName, false>`. 

I think you can use `std::optional<std::variant<bool, ESyncType>>` to do this:

https://github.com/multitheftauto/mtasa-blue/blob/0ee554b1cd5dc30c2c9ed348ab5ab94feb9e1ba4/Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp#L1577-L1585

Pirulax commented:
Alright. As it turned out, calling the `onElementDataChange` even takes the most time.
Calling the CCustomData::Set function barely takes time(6.8 us*), also all other function calls are pretty negligible, except this one(the whole process of calling `onElementDataChange`), this is why the `getData` is so much faster: it doesnt call events.  

so, time list(on x64 release server):
CStaticFuncDefs::SetElementData() 106 580 ns -> 106 us -> 0.1 ms, and 0.08 ms from this comes form calling the event, with only the default resources running. So it may take more, depending on the amount of resources, elements, due to the element tree size, as of my understanding.

This PR will help the client way more than the server, because the client is way busier than the server. 


*10 us on my CPU, take these measurements with a bag of salt, since im pretty sure they arent the most accurate due to the fact `CTimeMarkerQHD` is written: it needs to class CLOCK::now(), which translates to steady_clock::now(), which is already a small overhead + pushing the data back in the vector along with a stack allocated cstr, and everything, which makes every marker set after the first one inaccurate.

Pirulax commented:
> Take a look at CLuaBcryptDefs - you'll need to use `ArgumentParserWarn<MethodName, false>`.
> 
> I think you can use `std::optional<std::variant<bool, ESyncType>>` to do this:
> 
> https://github.com/multitheftauto/mtasa-blue/blob/0ee554b1cd5dc30c2c9ed348ab5ab94feb9e1ba4/Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp#L1577-L1585

Made a PR regarding that refactor, so I dont clutter this PR with non-necessary stuff anymore, cuz i did a few things not related to this PR at all.

Pirulax commented:
Alright, this PR adds no value at all(well, on client-side it does, but I haven't done that yet. I'll leave the server as it is)

Dutchman101 commented:
I don't agree, i feel it does add value and it would be a shame to let this effort go to waste.

In the case of elementdata, as long we're using it, there aren't many ways to optimize the implementation. I feel like you squeezed the most out of it with these changes, and even if the benefit is just a 20% performance gain, we should take it. It will already make a significant difference in today's climate of scripts (mis)using elementdata massively..

Pirulax commented:
The problem with this PR is that is incorporates a lot of changes, that aren't related to it at all(see refactors to other functions I think in CStaticFDs).  I'll make an other PR that'll actually be organized, and only contain changes related to it, because rn, reviewing it(let alone merging) is a PITA, IMHO.
